### DISCUSSION TITLE
Restrict data access for Users

### DISCUSSION THREAD

**Discussion by SantoshSrinivas79** [Budibase/budibase#3347] [Help] [answered]
If I understand right, the data access is based on Roles and everyone in the role gets access to the entire data in the table.

So, for example, if a user has access to the `Orders` table that is accessible for the `Basic` role with `Read & Write` permission.
All users in the `Basic` role can see and edit all the rows in that table.

This is obviously not desirable. Is there any way to restrict access within the same role for such a use case?

**Comment by mjashanks** [comment, marked as answer]
Roles currently work in a Hierarchal/Ordered way. So, for example, say I have roles in the order
1. Basic
2. Power
3. Super
4. Admin

On my `Orders` table, I could set 
- `Write : Super`
- `Read : Power`

This would mean
1. Basic has no access
2. Power has Read access
3. Super has Read/Write access
4. Admin has Read/Write access

You can create new roles as needed.

What we cannot do at the moment is create **per record permissions** - e.g. "Jane Smith has Full Access to Order # 43" - is this what you are asking for?

**Reply by SantoshSrinivas79** [reply]
Yep, correct. My use-case is **per record permissions**

However, a more mainstream use-case for business apps would perhaps be for example, "London Service Reps can service only London Customers"

**Reply by mjashanks** [reply]
Yes - I hear you. It might be possible to handle this scenario on the front-end.... by showing/hiding components, based on so properties of the "Current User".

However, this would still leave a security hole in the API, for users with the skill to use it.

It's something that I'd like us to support someday, but it won't be for a while.

**Reply by SantoshSrinivas79** [reply]
Yep! I did it in the front end but then hacked it via the API and saw the limitation. I'm thinking I will rely on REST endpoints to handle the security restrictions in the meantime. 

Does that make sense as a reasonable approach for now?

**Reply by mjashanks** [reply]
Yea - it does. we are planning on upgrading the REST connector in December as well FYI. If you are using it, feel free to post some feedback.
